include/RAJA/policy/cuda/reduce.hpp

```diff
@@ -109,1507 +190,1036 @@ __device__ __forceinline__ T shfl_xor(T var, int laneMask)
-   */
-  int m_smem_offset = -1;
+}  // end namespace cuda
 
-  /*!
-   * \brief If this variable is a copy or not; only original may release memory
-   *        or perform finalization.
-   */
-  bool m_is_copy_host = false;
-  bool m_is_copy_device = false;
-  bool m_finish_reduction = false;
-
-  // Sanity checks for block size and template type size
-  static constexpr bool powerOfTwoCheck = (!(BLOCK_SIZE & (BLOCK_SIZE - 1)));
-  static constexpr bool reasonableRangeCheck =
-      ((BLOCK_SIZE >= 32) && (BLOCK_SIZE <= 1024));
-  static constexpr bool sizeofcheck =
-      ((sizeof(T) <= sizeof(CudaReductionDummyDataType))
-       && (sizeof(CudaReductionTallyType<T>)
-           <= sizeof(CudaReductionDummyTallyType))
-       && (sizeof(CudaReductionBlockType<T>)
-           <= sizeof(CudaReductionDummyBlockType)));
-  static_assert(powerOfTwoCheck, "Error: block sizes must be a power of 2");
-  static_assert(reasonableRangeCheck,
-                "Error: block sizes must be between 32 and 1024");
-  static_assert(sizeofcheck,
-                "Error: type must be of size <= " RAJA_STRINGIFY_MACRO(
-                    RAJA_CUDA_REDUCE_VAR_MAXSIZE));
-};
+//! Cuda Target Reduction entity -- generalize on reduction, and type
+template <bool Async, typename Reducer, typename T>
+struct CudaReduce {
+  CudaReduce() = delete;
 
-/*!
- ******************************************************************************
- *
- * \brief  Sum reduction class template for use in CUDA kernel.
- *
- *         For usage example, see reducers.hpp.
- *
- ******************************************************************************
- */
-template <size_t BLOCK_SIZE, bool Async, typename T>
-class ReduceSum<cuda_reduce<BLOCK_SIZE, Async>, T>
-{
-public:
-  /*!
-   * \brief Constructor takes initial reduction value (default constructor
-   * is disabled).
-   *
-   * Note: Constructor only executes on the host.
-   */
-  explicit ReduceSum(T init_val)
+  explicit CudaReduce(T init_val)
+      : parent{this},
+        info{0},
+        val(init_val)
   {
-    m_is_copy_host = false;
-    m_myID = getCudaReductionId();
-    getCudaReductionMemBlock(m_myID, (void **)&m_blockdata);
-    getCudaReductionTallyBlock(m_myID,
-                               (void **)&m_tally_host,
-                               (void **)&m_tally_device);
-    m_tally_host->tally = init_val;
-    m_tally_host->retiredBlocks = static_cast<GridSizeType>(0);
   }
 
-  /*!
-   * \brief Initialize shared memory on device, request shared memory on host.
-   *
-   * Copy constructor executes on both host and device.
-   * On host requests dynamic shared memory and gets offset into dynamic
-   * shared memory if in forall.
-   * On device initializes dynamic shared memory to appropriate value.
-   */
-  __host__ __device__
-  ReduceSum(const ReduceSum<cuda_reduce<BLOCK_SIZE, Async>, T> &other)
+  RAJA_HOST_DEVICE
+  CudaReduce(const CudaReduce & other)
+#if !defined(__CUDA_ARCH__)
```

Comment:
I'm not sure if this is just my personal preference, but i'd rather have a macro guard around complete constructor implementations instead of conditionally setting single lines differently. Given that __CUDA_ARCH__ is checked multiple times, this might make it cleaner. Just my $0.02